File: LeetCodePractice/src/bst/ConstructBinaryTreeFromPreorderAndInorderTraversal.java
package bst;

import java.util.*;
/*
Given preorder and inorder traversal of a tree, construct the binary tree.

You may assume that duplicates do not exist in the tree.

Example
Given in-order [1,2,3] and pre-order [2,1,3], return a tree
*/
public class ConstructBinaryTreeFromPreorderAndInorderTraversal {	 
	  /**
	     *@param preorder : A list of integers that preorder traversal of a tree
	     *@param inorder : A list of integers that inorder traversal of a tree
	     *@return : Root of a tree
	     */
	    public TreeNode buildTree(int[] preorder, int[] inorder) {
	        // write your code here
	        if(preorder.length != inorder.length){
	            return null;
	        }
	        int len = inorder.length;
	        // in order value : index
	        HashMap<Integer, Integer> map = new HashMap<>();
	        for(int i = 0; i < len; i++){
	            map.put(inorder[i], i);
	        }
	        return helper(preorder, inorder, map, 0, len-1, 0, len-1);
	        
	    }
	    // need to index for partition
	    public TreeNode helper(int[] preorder, int[] inorder, HashMap<Integer, Integer> map, int preS, int preE, int inS, int inE){
	        if(preS > preE || inS > inE){
	            return null;
	        }
	        TreeNode root = new TreeNode(preorder[preS]);
	        int idx = map.get(root.val);
	        root.left = helper(preorder, inorder, map, preS + 1, preS + (idx - inS), inS, idx - 1);
	        root.right = helper(preorder, inorder, map, preS + (idx - inS) + 1, preE, idx + 1, inE);
	        return root;
	    }
}

File: LeetCodePython/ConstructBinaryTreeFromPreorderAndInorderTraversal.py
"""
Given two integer arrays preorder and inorder where preorder is the preorder traversal of a binary tree and inorder is the inorder traversal of the same tree, construct and return the binary tree.



Example 1:


Input: preorder = [3,9,20,15,7], inorder = [9,3,15,20,7]
Output: [3,9,20,null,null,15,7]
Example 2:

Input: preorder = [-1], inorder = [-1]
Output: [-1]


Constraints:

1 <= preorder.length <= 3000
inorder.length == preorder.length
-3000 <= preorder[i], inorder[i] <= 3000
preorder and inorder consist of unique values.
Each value of inorder also appears in preorder.
preorder is guaranteed to be the preorder traversal of the tree.
inorder is guaranteed to be the inorder traversal of the tree.
"""
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class ConstructBinaryTreeFromPreorderAndInorderTraversal:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        length = len(preorder)
        val_to_idx = {}
        for i, val in enumerate(inorder):
            val_to_idx[val] = i

        def helper(pre_s, pre_e, in_s, in_e):
            if pre_s > pre_e or in_s > in_e:
                return None
            root = TreeNode(preorder[pre_s])
            idx = val_to_idx[root.val]
            root.left = helper(pre_s + 1, pre_s + (idx - in_s), in_s, idx - 1)
            root.right = helper(pre_s + (idx - in_s) + 1, pre_e, idx + 1, in_e)
            return root

        return helper(0, length - 1, 0, length - 1)
